From b379d3f096dd3a4612d80fc923f99f518102d10b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bernhard=20=C3=9Cbelacker?= <bernhardu@mailbox.org>
Date: Tue, 10 Dec 2024 23:19:08 +0100
Subject: [PATCH 48/86] wsdapi: Use no_sanitize_address in find_allocation.

/home/bernhard/data/entwicklung/2024/wine/wine/tools/runtest -q -P wine -T . -M wsdapi.dll -p dlls/wsdapi/tests/x86_64-windows/wsdapi_test.exe memory && touch dlls/wsdapi/tests/x86_64-windows/memory.ok
=================================================================
==1160==ERROR: AddressSanitizer: heap-use-after-free on address 0x7f6035d00580 at pc 0x6ffff85f6e73 bp 0x7ffffe1ff860 sp 0x7ffffe1ff8a8
READ of size 4 at 0x7f6035d00580 thread T0
0494:fixme:file:server_get_file_info Unsupported info class e
    #0 0x6ffff85f6e72 in find_allocation /home/bernhard/data/entwicklung/2024/wine\wine/dlls/wsdapi/memory.c:58:21
    #1 0x6ffff85f6e72 in WSDAllocateLinkedMemory /home/bernhard/data/entwicklung/2024/wine\wine/dlls/wsdapi/memory.c:101:14
    #2 0x00014001d896 in AllocateLinkedMemory_tests /home/bernhard/data/entwicklung/2024/wine\wine/dlls/wsdapi/tests/memory.c:55:14
    #3 0x00014001d896 in func_memory /home/bernhard/data/entwicklung/2024/wine\wine/dlls/wsdapi/tests/memory.c:73:5
    #4 0x0001400375a0 in run_test /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:765:5
    #5 0x0001400375a0 in main /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:884:12
    #6 0x00014003940f in mainCRTStartup /home/bernhard/data/entwicklung/2024/wine\wine/dlls/msvcrt/crt_main.c:58:11
    #7 0x6ffffabf4808 in BaseThreadInitThunk /home/bernhard/data/entwicklung/2024/wine/wine/dlls/kernel32\thread.c:61:5
    #8 0x6ffffadcfa1a in RtlUserThreadStart (C:\windows\system32\ntdll.dll+0x17000fa1a)

0x7f6035d00580 is located 0 bytes inside of 1072-byte region [0x7f6035d00580,0x7f6035d009b0)
freed by thread T0 here:
    #0 0x6ffff885a1a1 in free /home/bernhard/data/entwicklung/2024/llvm-mingw/2024-10-18/llvm-mingw/llvm-project/compiler-rt\lib/asan/asan_malloc_win.cpp:71:3
    #1 0x00014001d889 in AllocateLinkedMemory_tests /home/bernhard/data/entwicklung/2024/wine\wine/dlls/wsdapi/tests/memory.c:52:5
    #2 0x00014001d889 in func_memory /home/bernhard/data/entwicklung/2024/wine\wine/dlls/wsdapi/tests/memory.c:73:5
    #3 0x0001400375a0 in run_test /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:765:5
    #4 0x0001400375a0 in main /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:884:12
    #5 0x00014003940f in mainCRTStartup /home/bernhard/data/entwicklung/2024/wine\wine/dlls/msvcrt/crt_main.c:58:11
    #6 0x6ffffabf4808 in BaseThreadInitThunk /home/bernhard/data/entwicklung/2024/wine/wine/dlls/kernel32\thread.c:61:5
    #7 0x6ffffadcfa1a in RtlUserThreadStart (C:\windows\system32\ntdll.dll+0x17000fa1a)

previously allocated by thread T0 here:
    #0 0x6ffff885a2c1 in malloc /home/bernhard/data/entwicklung/2024/llvm-mingw/2024-10-18/llvm-mingw/llvm-project/compiler-rt\lib/asan/asan_malloc_win.cpp:80:3
    #1 0x6ffff85f6d29 in WSDAllocateLinkedMemory /home/bernhard/data/entwicklung/2024/wine\wine/dlls/wsdapi/memory.c:88:11
    #2 0x00014001d724 in AllocateLinkedMemory_tests /home/bernhard/data/entwicklung/2024/wine\wine/dlls/wsdapi/tests/memory.c:46:14
    #3 0x00014001d724 in func_memory /home/bernhard/data/entwicklung/2024/wine\wine/dlls/wsdapi/tests/memory.c:73:5
    #4 0x0001400375a0 in run_test /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:765:5
    #5 0x0001400375a0 in main /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:884:12
    #6 0x00014003940f in mainCRTStartup /home/bernhard/data/entwicklung/2024/wine\wine/dlls/msvcrt/crt_main.c:58:11
    #7 0x6ffffabf4808 in BaseThreadInitThunk /home/bernhard/data/entwicklung/2024/wine/wine/dlls/kernel32\thread.c:61:5
    #8 0x6ffffadcfa1a in RtlUserThreadStart (C:\windows\system32\ntdll.dll+0x17000fa1a)

SUMMARY: AddressSanitizer: heap-use-after-free /home/bernhard/data/entwicklung/2024/wine\wine/dlls/wsdapi/memory.c:58:21 in find_allocation
Shadow bytes around the buggy address:
  0x7f6035d00300: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x7f6035d00380: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x7f6035d00400: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x7f6035d00480: fd fd fd fd fd fd fa fa fa fa fa fa fa fa fa fa
  0x7f6035d00500: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x7f6035d00580:[fd]fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x7f6035d00600: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x7f6035d00680: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x7f6035d00700: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x7f6035d00780: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x7f6035d00800: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==1160==ABORTING
0494:fixme:kernelbase:AppPolicyGetProcessTerminationMethod FFFFFFFFFFFFFFFA, 00007FFFFEA8FE80
make: *** [Makefile:457258: dlls/wsdapi/tests/x86_64-windows/memory.ok] Fehler 1
---
 dlls/wsdapi/memory.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/wsdapi/memory.c b/dlls/wsdapi/memory.c
index 7003f0aa4cb..5b7c4fdb413 100644
--- a/dlls/wsdapi/memory.c
+++ b/dlls/wsdapi/memory.c
@@ -44,6 +44,7 @@ struct memory_allocation
     struct list children;
 };
 
+__attribute__((no_sanitize_address))
 static struct memory_allocation *find_allocation(void *ptr)
 {
     struct memory_allocation *allocation;
-- 
2.47.1

